### ISSUE TITLE
Loss goes up with CrossBatchMemory?

### ISSUE THREAD

**Issue by JohnGiorgi** [KevinMusgrave/pytorch-metric-learning#24]
I am having a strange issue where my loss actually looks like its "flipped" when I wrap a loss function with [`CrossBatchMemory`](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#crossbatchmemory). My setup is a little contrived but here is the gist of what I call:

```python
from pytorch_metric_learning.losses import NTXentLoss, CrossBatchMemory

loss = NTXentLoss(0.1)
loss = CrossBatchMemory(loss, 128, memory_size=256)
```

When I _don't_ wrap with `CrossBatchMemory` loss, my loss goes down and my model trains without issue. When I do wrap it, my loss actually goes up:

<img width="369" alt="image" src="https://user-images.githubusercontent.com/8917831/76816054-9c86db00-67d5-11ea-99f8-2410891f1c8f.png">

Any ideas? Am I not using `CrossBatchMemory` as intended? Thanks!

**Comment by KevinMusgrave** [comment]
I think I found the bug: The paper says to add the current batch to the memory queue before computing the loss. I'm not doing that. Instead I'm adding it to the memory queue **after** the loss is computed. So at each iteration, the loss is calculated without considering any of the relationships within the current batch. It only considers relationships between the current batch and the memory queue. The reason I didn't notice this bug before was because I was using a large memory_size, like in the thousands, and it seemed to be working. But I tried 256 like you and it totally fails.

I also found a .to(device) bug in CrossBatchMemory, but that's unrelated to the loss bug you're seeing.

Here are the changes in [CrossBatchMemory](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/pytorch_metric_learning/losses/cross_batch_memory.py#L18) that worked for me:
```python
        labels = labels.to(embeddings.device) ### this addresses the .to(device) bug
        self.embedding_memory = self.embedding_memory.to(embeddings.device)
        self.label_memory = self.label_memory.to(labels.device)
        self.add_to_memory(embeddings, labels, batch_size) ### this addresses the loss bug
        ...
        loss = self.loss(combined_embeddings, combined_labels, indices_tuple)
        self.add_to_memory(embeddings, labels, batch_size) ### delete this
        return loss
```

Larger memory_size still seems to help.

I'll try to commit and upload soon
